toolchain/check/inst.h

```diff
@@ -19,15 +21,17 @@ auto AddInst(Context& context, SemIR::LocIdAndInst loc_id_and_inst)
 //
 // As a safety check, prevent use with storage insts (see `AddInstWithCleanup`).
 template <typename InstT, typename LocT>
-  requires(!InstT::Kind.has_cleanup())
+  requires(!InstT::Kind.has_cleanup() &&
+           std::convertible_to<LocT, SemIR::LocId>)
```

Comment:
I disagree with all the explicit construction, but it seems I'm the minority vote. Given that, why isn't every `AddInst` call requiring an explicit `LocId` instead of `LocT`?

Reply:
i.e., down this path, I think there should be no use of `std::convertible_to` in any of these files.

Reply:
The reason is because `LocIdAndInst` constructor needs to know the original type for its concept checks on the NodeId, so we can't throw that away and convert to LocId earlier

Reply:
Did the PR not work without these `requires` though? To put this concretely, if I check out trunk and revert the changes to this file, the toolchain builds and passes tests. It feels a lot to me like you're adding these because you want different build errors; if so, it seems they aren't necessary to do this PR ("Construct LocId from InstId directly"). I was also getting confused because these aren't explained in the PR description, and I was making incorrect assumptions about why changes would be necessary here.

Am I correct that these are just being added to get different build errors? If so, can they be removed from this PR and split out?

Reply:
Ah, yes this moves the build errors when you do it wrong to the caller instead of inside the constructor this this function calls. I will add to the PR description.